scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java

```diff
@@ -0,0 +1,388 @@
+
+    final ValuePathExpression valuePathExpression = valuePathExpression(patchOperation);
+    final AttributeReference attributeReference = attributeReference(valuePathExpression);
+
+    PatchOperationHandler patchOperationHandler = patchOperationHandlers.get(patchOperation.getOperation());
+
+    // if the attribute has a URN, assume it's an extension that URN does not match the baseUrn
+    if (attributeReference.hasUrn() && !attributeReference.getUrn().equals(source.getBaseUrn())) {
+      Schema schema = this.schemaRegistry.getSchema(attributeReference.getUrn());
+      Attribute attribute = schema.getAttribute(attributeReference.getAttributeName());
+      checkMutability(schema.getAttributeFromPath(attributeReference.getFullAttributeName()));
+
+      patchOperationHandler.applyExtensionValue(source, sourceAsMap, schema, attribute, valuePathExpression, attributeReference.getUrn(), patchOperation.getValue());
+    } else {
+      Schema schema = this.schemaRegistry.getSchema(source.getBaseUrn());
+      Attribute attribute = schema.getAttribute(attributeReference.getAttributeName());
+      checkMutability(schema.getAttributeFromPath(attributeReference.getFullAttributeName()));
+
+      patchOperationHandler.applyValue(source, sourceAsMap, schema, attribute, valuePathExpression, patchOperation.getValue());
+    }
+  }
+
+  private PatchOperationPath tryGetOperationPath(String key) {
+    try {
+      return new PatchOperationPath(key);
+    } catch (FilterParseException e) {
+      log.warn("Parsing path failed with exception.", e);
+      throw new IllegalArgumentException("Cannot parse path expression: " + e.getMessage());
+    }
+  }
+
+  private Map<String, Object> objectAsMap(final Object object) {
+    return objectMapper.convertValue(object, MAP_TYPE);
+  }
+
+  public static ValuePathExpression valuePathExpression(final PatchOperation operation) {
+    return Optional.ofNullable(operation.getPath())
+      .map(PatchOperationPath::getValuePathExpression)
+      .orElseThrow(() -> new IllegalArgumentException("Patch operation must have a value path expression"));
+  }
+
+  public static AttributeReference attributeReference(final ValuePathExpression expression) {
+    return Optional.ofNullable(expression.getAttributePath())
+      .orElseThrow(() -> new IllegalArgumentException("Patch operation must have an expression with a valid attribute path"));
+  }
+
+  private static void checkMutability(Attribute attribute) throws IllegalArgumentException {
+    if (attribute.getMutability().equals(Attribute.Mutability.READ_ONLY)) {
+      String message = "Can not update a read-only attribute '" + attribute.getName() + "'";
+      log.error(message);
+      throw new IllegalArgumentException(message);
+    }
+  }
+
+  private static void checkMutability(Attribute attribute, Object currentValue) throws IllegalArgumentException {
```

Comment:
My last change litters the code with a few more `checkMutability` calls, to account for rfc7644 sec 3.5.2

> a client MUST NOT modify an attribute that has mutability "readOnly" or "immutable".  However, a client MAY "add" a value to an "immutable" attribute if the attribute had no previous value.

I'm not really a fan of spreading of how I spread this logic around, but needing the object's current value makes it tricky.
I went down the rabbit hole of creating a Collection/Map/Iterator/etc that could wrap a delegate and then make a mutability check if/when the attribute was mutated in the map.  It was a neat bit of code, but it was ugly and required special handling for extensions so I threw it away.  There is probably something more elegant that could be done, but I'm out of ideas 😆 

It's not critical to get this in, but if you have any thoughts let me know!